gammapy/maps/hpxnd.py

```diff
@@ -410,6 +415,16 @@ def fill_by_idx(self, idx, weights=None):
         msk = idx_local[0] >= 0
         idx_local = [t[msk] for t in idx_local]
         if weights is not None:
+            if isinstance(weights, Quantity):
+                if self.unit.is_equivalent(weights.unit):
+                    weights = np.asarray(weights.to(self.unit).value, dtype=self.data.dtype)
```

Comment:
I think we should try to simplify these checks and handling for weights.

@registerrier - Why are you calling `np.asarray` and forcing a `dtype`?
That makes a copy, i.e. is inefficient.
There is no reason to introduce a copy and dtype change for weights, no?

Another thing I note is that these 10 complex lines of type checking are duplicated for HPX and WCS.

@registerrier - Can you try to remove the duplication (few line utility function?) and also try to simplify the logic to have less code here?

I haven't thought about it much, but shouldn't Astropy do the quantity handling for you and through an error if they don't match? Why do you have to do quantity -> array -> quantity and raise an error yourself here?

Reply:
I simply reused here the code as initially written in `gammapy.maps`, to ensure stable behavior at least for empty units. I think for the sake of stability I would leave as is for the moment.
Note BTW that from `np.asarray` documentation, no copy is performed if the input is already an ndarray with matching dtype and order.  So the inefficiency is likely minimal.

Regarding the quantity-array conversion. The code is making some check for multiple entries to fill in a given bin. Hence the `np.bincount`, later in the code. I don"t know if it works well with weights  of type `Quantity`. 

Again if we are to make rapid changes to have 3D working, we should postpone this type of improvement for later, no?